File: project.py
from decimal import Decimal, getcontext
import math
from typing import Callable, List
getcontext().prec = 110 # this allows for 110 decimal places when doing the sqrt operation

# bad
def subset_sum_backwards(M: list[int], S: int) -> list[int]:
    """
    M is a list (length 100) of integers and S is the target sum.
    Return the subset that sums the closest to S.
    """
    T = [0] * len(M) # binary list to show if the element from M is in the subset

    # Start at the end of the list and work backwards
    # only add elements to the subset if they don't make the sum exceed S
    for i in range(len(M) - 1, -1, -1):
        if sum(M[j] for j in range(len(M)) if T[j]) + M[i] <= S:
            T[i] = 1 # include this element in the subset

    return T

# bad
def subset_sum_forwards(M: list[int], S: int) -> list[int]:
    """
    M is a list (length 100) of integers and S is the target sum.
    Return the subset that sums the closest to S.
    """
    T = [0] * len(M) # binary list to show if the element from M is in the subset

    # Start at the beginning of the list and work forwards
    # only add elements to the subset if they don't make the sum exceed S
    for i in range(len(M)):
        if sum(M[j] for j in range(len(M)) if T[j]) + M[i] <= S:
            T[i] = 1 # include this element in the subset

    return T

# worse
def subset_sum_alternate(M: list[int], S: int) -> list[int]:
    """
    M is a list (length 100) of integers and S is the target sum.
    Return the subset that sums the closest to S.
    """
    T = [0] * len(M) # binary list to show if the element from M is in the subset

    # Choose the first ele, then the last, then the second, then the second to last, etc.
    # This is a bit of a hack, but it works for this problem.
    for i in range(len(M) // 2):
        if sum(M[j] for j in range(len(M)) if T[j]) + M[i] <= S:
            T[i] = 1
        if sum(M[j] for j in range(len(M)) if T[j]) + M[len(M) - 1 - i] <= S:
            T[len(M) - 1 - i] = 1

    return T

# too slow
def subset_sum_brute_force(M: list[int], S: int) -> list[int]:
    """
    M is a list (length 100) of integers and S is the target sum.
    Return the subset that sums the closest to S.
    """
    T = [0] * len(M) # binary list to show if the element from M is in the subset

    # Check all possible subsets
    print(f"Checking {1 << len(M)} possible subsets...")
    for i in range(1 << len(M)):
        if i % 1000000 == 0:
            print(f"Checking subset {i}...")
        current_sum = sum(M[j] for j in range(len(M)) if (i & (1 << j)) > 0)
        if current_sum <= S:
            T = [(i & (1 << j)) > 0 for j in range(len(M))]

    return T

# worse
def subset_sum_recursive(M: list[int], S: int, index: int, T: list[int]) -> list[int]:
    """
    M is a list (length 100) of integers and S is the target sum.
    Return the subset that sums the closest to S.
    """
    if index == len(M):
        return T

    # Include the current element in the subset
    T[index] = 1
    if sum_from_bin_list(T, M) <= S:
        return subset_sum_recursive(M, S, index + 1, T)

    # Exclude the current element from the subset
    T[index] = 0
    return subset_sum_recursive(M, S, index + 1, T)

def subset_sum_refine(M: list[int], S: int, subset_sum_function: Callable) -> list[int]:
    """
    M is a list (length 100) of integers and S is the target sum.
    Return the subset that sums the closest to S.
    Calls a different function to start this process, then refines the result.
    """

    T0 = subset_sum_function(M, S)
    sum0 = sum_from_bin_list(T0, M)
    err = abs(S - sum0)
    # Refine the result with the hill climbing algorithm
    improve = True
    while improve:
        improve = False
        # foreach M[i] == 1: remove it and try to add M[j] == 0
        for i in range(len(M)):
            if T0[i] == 1:
                for j in range(len(M)):
                    if T0[j] == 0:
                        # remove M[i] and add M[j]
                        new_sum = sum0 - M[i] + M[j]
                        new_err = abs(S - new_sum)
                        if new_err < err:
                            # update the subset and the error
                            print(f"Improving: removing M[{i}] = {M[i]} and adding M[{j}] = {M[j]}")
                            T0[i] = 0
                            T0[j] = 1
                            sum0 = new_sum
                            err = new_err
                            improve = True
                            break
                if improve:
                    break

    return T0

def subset_sum_refine_2(M: list[int], S: int, subset_sum_function: Callable, k=2, l=3) -> list[int]:
    """
    M is a list (length 100) of integers and S is the target sum.
    Return the subset that sums the closest to S.
    Calls a different function to start this process, then refines the result.
    k: Number of elements to remove on each iteration.
    l: Number of elements to add back on each iteration.
    """

    print(f"Refining with k = {k} and l = {l}")
    T0 = subset_sum_function(M, S)
    sum0 = sum_from_bin_list(T0, M)
    err = abs(S - sum0)
    # Refine the result with the hill climbing algorithm
    improve = True
    while improve:
        improve = False
        # Generate all combinations of k elements to remove
        remove_indices = [i for i in range(len(M)) if T0[i] == 1]
        add_indices = [i for i in range(len(M)) if T0[i] == 0]

        from itertools import combinations

        for remove_comb in combinations(remove_indices, k):
            for add_comb in combinations(add_indices, l):
                # Calculate the new sum after removing and adding elements
                new_sum = sum0
                for r in remove_comb:
                    new_sum -= M[r]
                for a in add_comb:
                    new_sum += M[a]

                new_err = abs(S - new_sum)
                if new_err < err:
                    # Update the subset and the error
                    print(f"Improving: removing {remove_comb} and adding {add_comb}")
                    for r in remove_comb:
                        T0[r] = 0
                    for a in add_comb:
                        T0[a] = 1
                    sum0 = new_sum
                    err = new_err
                    improve = True
                    break
            if improve:
                break

    return T0

def subset_sum_refine_3(M: list[int], S: int, subset_sum_function: Callable, k_l_pairs: List[tuple], N=None) -> list[int]:
    """
    M is a list (length 100) of integers and S is the target sum.
    Return the subset that sums the closest to S.
    Calls a different function to start this process, then refines the result.
    k_l_pairs: A list of (k, l) pairs to use for each iteration.
    N: Number of iterations (optional). If not provided, defaults to the length of k_l_pairs.
    """

    if N is None:
        N = len(k_l_pairs)

    print(f"Refining with {N} iterations and varying k, l pairs", flush=True)
    T0 = subset_sum_function(M, S)
    sum0 = sum_from_bin_list(T0, M)
    err = abs(S - sum0)

    for i in range(N):
        k, l = k_l_pairs[i % len(k_l_pairs)]  # Cycle through k_l_pairs if N > len(k_l_pairs)
        print(f"Iteration {i + 1} of {N}: Using k = {k}, l = {l}", flush=True)
        improve = True
        while improve:
            improve = False
            # Generate all combinations of k elements to remove
            remove_indices = [i for i in range(len(M)) if T0[i] == 1]
            add_indices = [i for i in range(len(M)) if T0[i] == 0]

            from itertools import combinations

            for remove_comb in combinations(remove_indices, k):
                for add_comb in combinations(add_indices, l):
                    # Calculate the new sum after removing and adding elements
                    new_sum = sum0
                    for r in remove_comb:
                        new_sum -= M[r]
                    for a in add_comb:
                        new_sum += M[a]

                    new_err = abs(S - new_sum)
                    if new_err < err:
                        # Update the subset and the error
                        print(f"Improving: removing {remove_comb} and adding {add_comb}", flush=True)
                        for r in remove_comb:
                            T0[r] = 0
                        for a in add_comb:
                            T0[a] = 1
                        sum0 = new_sum
                        err = new_err
                        improve = True
                        break
                if improve:
                    break

    return T0

def dp_subset_sum(M, S):
    dp = {0: []}  # Dictionary to store sums and their corresponding subsets
    for num in M:
        for current_sum, subset in list(dp.items()):
            new_sum = current_sum + num
            if new_sum <= S and new_sum not in dp:
                dp[new_sum] = subset + [num]
    closest_sum = max(dp.keys())
    return dp[closest_sum]

def subset_sum_split_recursive(
    M: list[int], 
    S: int, 
    depth: int = 0, 
    subset_function: Callable[[list[int], int], list[int]] = dp_subset_sum,
    max_depth: int = 2,
    num_splits: int = 2,
    corrector_function: Callable[[list[int], int, Callable], list[int]] = subset_sum_refine
) -> list[int]:
    """
    Recursively split the list M into smaller subsets down to a specified depth.
    At the base case, apply the provided subset_function to calculate the subset.
    After combining the results, apply a corrector function to refine the subset.

    Args:
        M: List of integers.
        S: Target sum.
        depth: Depth of recursion.
        subset_function: Function to calculate the subset for smaller lists.
        max_depth: Maximum depth of recursion.
        num_splits: Number of splits at each level.
        corrector_function: Function to refine the combined subset.

    Returns:
        A binary list indicating the subset that sums closest to S.
    """
    print(f"Depth: {depth}, M: {M}, S: {S}")

    if not M:
        return []
    if len(M) == 1:
        return [1] if M[0] <= S else [0]
    if depth == max_depth:
        return subset_function(M, S)

    # Split M into `num_splits` parts
    chunk_size = (len(M) + num_splits - 1) // num_splits  # Calculate chunk size
    chunks = [M[i * chunk_size:(i + 1) * chunk_size] for i in range(num_splits)]

    # Split the target sum approximately evenly among the chunks
    target_splits = [S // num_splits] * num_splits
    for i in range(S % num_splits):
        target_splits[i] += 1  # Distribute the remainder

    # Recursively calculate subsets for each chunk
    subsets = [
        subset_sum_split_recursive(chunks[i], target_splits[i], depth + 1, subset_function, max_depth, num_splits, corrector_function)
        for i in range(len(chunks))
    ]

    # Combine the results into a single binary list
    T = [0] * len(M)
    offset = 0
    for i, subset in enumerate(subsets):
        for j, value in enumerate(subset):
            T[offset + j] = value
        offset += len(chunks[i])

    # Apply the corrector function to refine the combined subset
    T = corrector_function(M, S, lambda m, s: T)

    return T
    

def is_prime(n: int) -> bool:
    """
    Check if n is prime.
    """
    if n < 2:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

def sum_from_bin_list(T: list[int], M: list[int]) -> int:
    """
    Given a binary list T and a list M, return the sum of the elements in M
    that are included in the subset indicated by T.
    """
    return sum(M[i] for i in range(len(M)) if T[i])

def timer(func: Callable) -> Callable:
    """
    Decorator to time a function.
    """
    def wrapper(*args, **kwargs):
        import time
        start_time = time.time()
        result = func(*args, **kwargs)
        end_time = time.time()
        print(f"Time taken: {end_time - start_time:.2f} seconds")
        return result
    return wrapper

@timer
def main(subset_sum_function=subset_sum_backwards, ID=113596441, extra_args=[]):
    print(f"Using {subset_sum_function.__name__} with ID {ID}")
    print()
    # generate the first 100 prime numbers
    primes = []
    i = 2
    while len(primes) < 100:
        if is_prime(i):
            primes.append(i)
        i += 1

    # print(len(primes))
    # print(primes)
    # print()

    # M = [10**100 * i**0.5 for i in primes]
    M = [math.floor(Decimal(i).sqrt() * Decimal(10**100)) for i in primes]
    # print(len(M))
    # print(M)

    S = 7 * ID * 10**94

    print("Number of digits in the target sum:", len(str(int(S))))
    print("Target sum:", S)
    print()

    exp_target = sum_from_bin_list(subset_sum_function(M, S, *extra_args), M)
    print("Number of digits in the subset sum:", len(str(int(exp_target))))
    print("Subset sum:", exp_target)
    print()

    print("Number of digits in the difference:", len(str(int(S - exp_target))))
    print("Difference:", S - exp_target) 
    print()


ID = 113596441
# main(subset_sum_function=subset_sum_backwards, ID=ID, extra_args=[])
# main(subset_sum_function=subset_sum_forwards, ID=ID, extra_args=[])
# main(subset_sum_function=subset_sum_alternate, ID=ID, extra_args=[])
# main(subset_sum_function=subset_sum_brute_force, ID=ID, extra_args=[])
# main(subset_sum_function=subset_sum_recursive, ID=ID, extra_args=[0, [0] * 100])
# main(subset_sum_function=subset_sum_refine, ID=ID, extra_args=[subset_sum_backwards])


# subset_sum_refine_2 with subset_sum_backwards as start and k = 2, l = 3, makes the error 94 digits 
# main(subset_sum_function=subset_sum_refine_2, ID=ID, extra_args=[subset_sum_backwards, 2, 3])

# subset_sum_refine_3 with subset_sum_backwards as start and k = 2, l = 3, N = 3 makes the error 94 digits 
main(subset_sum_function=subset_sum_refine_3, ID=ID, extra_args=[subset_sum_alternate, [(1, 1), (1, 2), (2, 3), (3, 3), (3, 2), (2, 1), (1, 1)]])


# main(subset_sum_function=subset_sum_split_recursive, ID=ID, extra_args=[subset_sum_backwards, 2, 3, subset_sum_refine_3, [(1, 1), (1, 2), (2, 3)], 3])

File: project.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <numeric>
#include <chrono>
#include <iomanip>
#include <random>
#include <string>
#include <functional> // Add this header for std::function

using namespace std;

// Calculate sum from binary list
double sum_from_bin_list(const vector<int>& T, const vector<double>& M) {
    double sum = 0;
    for (size_t i = 0; i < T.size(); ++i) {
        if (T[i]) {
            sum += M[i];
        }
    }
    return sum;
}

// Function to generate combinations
void generate_combinations(vector<int>& current, int start, int n, int k, 
                          vector<vector<int>>& result, const vector<int>& indices) {
    if (k == 0) {
        result.push_back(current);
        return;
    }
    
    for (int i = start; i <= n - k; ++i) {
        current.push_back(indices[i]);
        generate_combinations(current, i + 1, n, k - 1, result, indices);
        current.pop_back();
    }
}

// Function to display progress bar
void display_progress_bar(long step, long total_steps, int barWidth = 40) {
    int pct = int((step * 100) / total_steps);
    int pos = barWidth * pct / 100;
    
    cout << "\r[";
    for (int b = 0; b < barWidth; ++b) {
        if (b < pos) cout << "=";
        else if (b == pos) cout << ">";
        else cout << " ";
    }
    cout << "] " << setw(3) << pct << "% "
         << "(" << step << "/" << total_steps << ")"
         << flush;
}

// Modified to accept std::function instead of function pointer
vector<int> subset_sum_progressive(const vector<double>& M, 
                                double S, 
                                function<vector<int>(const vector<double>&, double)> subset_sum_function) {
    // Initial subset
    vector<int> T0 = subset_sum_function(M, S);
    double sum0 = sum_from_bin_list(T0, M);
    double err = fabs(S - sum0);
    
    cout << "Initial solution error: " << scientific << err << endl;
    cout << "Target sum: " << S << endl;
    cout << "Initial sum: " << sum0 << endl;
    
    // Progressive refinement strategy with different phases
    vector<vector<pair<int, int>>> phases = {
        // Phase 1: Fine-tuning with small swaps
        {{1, 1}, {1, 2}, {2, 1}},
        
        // Phase 2: Medium adjustments
        {{2, 2}, {2, 3}, {3, 2}},
        
        // Phase 3: Larger structural changes
        {{3, 3}, {3, 4}, {4, 3}},

        // Phase 4: Larger structural changes
        {{4, 4}, {4, 5}, {5, 4}},
        
        // Phase 5: Final adjustments
        {{1, 1}, {1, 2}, {2, 1}}
    };
    
    for (size_t phase = 0; phase < phases.size(); phase++) {
        cout << "\n=== Phase " << phase + 1 << " ===\n";
        
        for (const auto& k_l_pair : phases[phase]) {
            int k = k_l_pair.first;
            int l = k_l_pair.second;
            
            cout << "\nTrying k=" << k << ", l=" << l << endl;
            
            bool improve = true;
            int iteration = 0;
            
            while (improve && iteration < 5) { // Limit iterations per k,l pair
                iteration++;
                improve = false;
                
                // Build remove/add index lists
                vector<int> remove_indices, add_indices;
                for (size_t i = 0; i < T0.size(); ++i) {
                    if (T0[i]) {
                        remove_indices.push_back(i);
                    } else {
                        add_indices.push_back(i);
                    }
                }
                
                if (k > (int)remove_indices.size() || l > (int)add_indices.size()) {
                    cout << "Skipping: not enough elements to remove/add\n";
                    break;
                }
                
                // Precompute all combinations
                vector<vector<int>> remove_combinations, add_combinations;
                vector<int> tmp;
                generate_combinations(tmp, 0, remove_indices.size(), k, 
                                     remove_combinations, remove_indices);
                tmp.clear();
                generate_combinations(tmp, 0, add_indices.size(), l, 
                                     add_combinations, add_indices);
                
                const long total_remove = remove_combinations.size();
                const long total_add = add_combinations.size();
                const long total_steps = total_remove * total_add;
                
                cout << "Checking " << total_steps << " combinations...\n";
                
                // Track best improvement
                double best_err = err;
                vector<int> best_remove, best_add;
                
                long step = 0;
                for (int ri = 0; ri < (int)total_remove && !improve; ++ri) {
                    for (int ai = 0; ai < (int)total_add; ++ai) {
                        ++step;
                        
                        // Update progress bar every 1000 steps or for last step
                        if (step % 1000 == 0 || step == total_steps) {
                            display_progress_bar(step, total_steps);
                        }
                        
                        // Test this remove/add pair
                        double new_sum = sum0;
                        for (int r : remove_combinations[ri]) {
                            new_sum -= M[r];
                        }
                        for (int a : add_combinations[ai]) {
                            new_sum += M[a];
                        }
                        
                        double new_err = fabs(S - new_sum);
                        
                        // Check if this is better than our best so far
                        if (new_err < best_err) {
                            best_err = new_err;
                            best_remove = remove_combinations[ri];
                            best_add = add_combinations[ai];
                            
                            // If it's significantly better, stop early
                            if (best_err < err * 0.5) {
                                improve = true;
                                break;
                            }
                        }
                    }
                }
                
                cout << endl; // End the progress bar line
                
                // Apply the best improvement found (if any)
                if (best_err < err) {
                    // Clear progress bar line
                    cout << "\rImprovement found: " << scientific << err << " → " << best_err << endl;
                    
                    // Apply changes
                    for (int r : best_remove) {
                        T0[r] = 0;
                    }
                    for (int a : best_add) {
                        T0[a] = 1;
                    }
                    
                    sum0 = sum_from_bin_list(T0, M);  // Recalculate for numerical stability
                    err = best_err;
                    improve = true;
                } else {
                    cout << "No improvement found for k=" << k << ", l=" << l << endl;
                }
            }
        }
    }
    
    return T0;
}

// Original subset sum function (backwards greedy approach)
vector<int> subset_sum_backwards(const vector<double>& M, double S) {
    vector<int> T(M.size(), 0);
    double current_sum = 0;
    
    for (int i = M.size() - 1; i >= 0; --i) {
        if (current_sum + M[i] <= S) {
            T[i] = 1;
            current_sum += M[i];
        }
    }
    
    return T;
}

// Forward greedy approach (alternative starting point)
vector<int> subset_sum_forwards(const vector<double>& M, double S) {
    vector<int> T(M.size(), 0);
    double current_sum = 0;
    
    for (int i = 0; i < M.size(); ++i) {
        if (current_sum + M[i] <= S) {
            T[i] = 1;
            current_sum += M[i];
        }
    }
    
    return T;
}

// Value ratio based approach (tries to maximize value per element)
vector<int> subset_sum_value_ratio(const vector<double>& M, double S) {
    vector<pair<double, int>> value_ratio;
    for (int i = 0; i < M.size(); i++) {
        value_ratio.push_back({M[i], i});
    }
    
    // Sort by value (ascending)
    sort(value_ratio.begin(), value_ratio.end());
    
    vector<int> T(M.size(), 0);
    double current_sum = 0;
    
    // Try to add items in order of increasing value
    for (const auto& item : value_ratio) {
        if (current_sum + M[item.second] <= S) {
            T[item.second] = 1;
            current_sum += M[item.second];
        }
    }
    
    return T;
}

int main() {
    // Generate first 100 primes
    vector<int> primes;
    int num = 2;
    while (primes.size() < 100) {
        bool is_prime = true;
        for (int i = 2; i * i <= num; ++i) {
            if (num % i == 0) {
                is_prime = false;
                break;
            }
        }
        if (is_prime) {
            primes.push_back(num);
        }
        num++;
    }
    
    // Create M = {sqrt(p_i) | 1 <= i <= 100 where p_i is the ith prime}
    vector<double> M(100);
    for (int i = 0; i < 100; ++i) {
        M[i] = sqrt(static_cast<double>(primes[i]));
    }
    
    // Target sum without 10^100 factor
    double S = 7 * 113596441 * pow(10, -6);
    
    cout << "=================================================\n";
    cout << "Subset Sum Progressive Refinement Algorithm\n";
    cout << "=================================================\n";
    cout << "Target sum: " << S << "\n";
    cout << "Number of elements: " << M.size() << "\n\n";
    
    auto start = chrono::high_resolution_clock::now();
    
    // Try multiple starting points and choose the best result
    cout << "Testing different starting algorithms...\n\n";
    
    vector<vector<int>> starting_solutions = {
        subset_sum_backwards(M, S),
        subset_sum_forwards(M, S),
        subset_sum_value_ratio(M, S)
    };
    
    vector<string> algorithm_names = {
        "Backwards Greedy",
        "Forwards Greedy",
        "Value Ratio Approach"
    };
    
    vector<int> best_solution;
    double best_error = numeric_limits<double>::max();
    int best_algo_index = -1;
    
    // Find the best starting solution
    for (size_t i = 0; i < starting_solutions.size(); i++) {
        double sum = sum_from_bin_list(starting_solutions[i], M);
        double error = fabs(S - sum);
        
        cout << algorithm_names[i] << ":\n";
        cout << "  Sum: " << sum << "\n";
        cout << "  Error: " << scientific << error << "\n\n";
        
        if (error < best_error) {
            best_error = error;
            best_solution = starting_solutions[i];
            best_algo_index = i;
        }
    }
    
    cout << "Using best starting solution: " << algorithm_names[best_algo_index] << "\n";
    cout << "Initial error: " << scientific << best_error << "\n\n";
    
    // Apply progressive refinement to the best starting solution
    vector<int> result = subset_sum_progressive(M, S, 
        [&best_solution](const vector<double>& M, double S) {
            return best_solution;
        });
    
    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> duration = end - start;
    
    cout << "\n=================================================\n";
    cout << "Results:\n";
    cout << "=================================================\n";
    
    double final_sum = sum_from_bin_list(result, M);
    double final_error = fabs(S - final_sum);
    
    cout << "Elements in subset: " << count(result.begin(), result.end(), 1) << "/" << M.size() << endl;
    cout << "Final sum (unscaled): " << final_sum << endl;
    cout << "Target sum (unscaled): " << S << endl;
    cout << "Absolute difference: " << final_error << endl;
    cout << "Relative error: " << (final_error / S) * 100 << "%" << endl;

    // Adding scaling factor back for final report
    cout << "\nWith 10^100 scaling factor:" << endl;
    cout << "Final sum: " << final_sum << " × 10^100" << endl;
    cout << "Target sum: " << S << " × 10^100" << endl;
    cout << "Difference: " << final_error << " × 10^100" << endl;
        
    // Calculate digit precision considering the 10^100 scaling factor
    int precision_digits = 0;
    if (final_error > 0) {
        // For a number scaled by 10^100, we need to consider how many digits after
        // the most significant digit are correctly matched
        precision_digits = static_cast<int>(100 + log10(final_error));
    }
        
    cout << "Matched to approximately " << precision_digits << " digits of precision" << endl;
    cout << "Time taken: " << duration.count() << " seconds" << endl;
    
    // Display elements in the final subset
    cout << "\nFinal subset elements:" << endl;
    for (size_t i = 0; i < result.size(); ++i) {
        if (result[i]) {
            cout << "  M[" << i << "] = " << M[i] << " (sqrt of " << primes[i] << ")" << endl;
        }
    }
    
    return 0;
}
